File: backend-python/app/repositories/feedback_repository.py
"""
Feedback Repository
Database operations for feedback and NPS entities
"""

import json
import logging
from typing import Any
from uuid import uuid4

from app.database import fetch, fetchone, execute, record_to_dict, records_to_list
from app.models.feedback import FeedbackData, CreateFeedbackRequest, UpdateFeedbackRequest, NPSData

logger = logging.getLogger(__name__)


class FeedbackRepository:
    """Repository for feedback database operations."""

    async def find_by_id(self, feedback_id: str) -> FeedbackData | None:
        """Find feedback by ID from nps_feedback table."""
        try:
            query = """
                SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
                FROM nps_feedback WHERE id = $1
            """
            record = await fetchone(query, feedback_id)
            if not record:
                return None
            data = record_to_dict(record)
            # Convert to FeedbackData format
            return FeedbackData(
                id=data["id"],
                simulation_id=data["simulation_id"],
                user_id=data["user_id"],
                competency_id=None,
                rating=data.get("score", 0) * 10,  # Convert 0-10 to 0-100
                comments=data.get("comments"),
                feedback_type=data["feedback_type"],
                metadata=None,
                created_at=data.get("submitted_at"),
                updated_at=data.get("submitted_at"),
            )
        except Exception as e:
            logger.error(f"Error finding feedback by ID: {e}")
            raise

    async def find_by_simulation_id(self, simulation_id: str) -> list[FeedbackData]:
        """Find all feedback for a simulation from nps_feedback table."""
        try:
            query = """
                SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
                FROM nps_feedback WHERE simulation_id = $1
                ORDER BY submitted_at DESC
            """
            records = await fetch(query, simulation_id)
            feedbacks = []
            for record in records:
                data = record_to_dict(record)
                feedbacks.append(FeedbackData(
                    id=data["id"],
                    simulation_id=data["simulation_id"],
                    user_id=data["user_id"],
                    competency_id=None,
                    rating=data.get("score", 0) * 10,
                    comments=data.get("comments"),
                    feedback_type=data["feedback_type"],
                    metadata=None,
                    created_at=data.get("submitted_at"),
                    updated_at=data.get("submitted_at"),
                ))
            return feedbacks
        except Exception as e:
            logger.error(f"Error finding feedback by simulation ID: {e}")
            raise

    async def find_by_user_id(self, user_id: str) -> list[FeedbackData]:
        """Find all feedback for a user from nps_feedback table."""
        try:
            query = """
                SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
                FROM nps_feedback WHERE user_id = $1
                ORDER BY submitted_at DESC
            """
            records = await fetch(query, user_id)
            feedbacks = []
            for record in records:
                data = record_to_dict(record)
                feedbacks.append(FeedbackData(
                    id=data["id"],
                    simulation_id=data["simulation_id"],
                    user_id=data["user_id"],
                    competency_id=None,
                    rating=data.get("score", 0) * 10,
                    comments=data.get("comments"),
                    feedback_type=data["feedback_type"],
                    metadata=None,
                    created_at=data.get("submitted_at"),
                    updated_at=data.get("submitted_at"),
                ))
            return feedbacks
        except Exception as e:
            logger.error(f"Error finding feedback by user ID: {e}")
            raise

    async def create(self, feedback_data: CreateFeedbackRequest | dict[str, Any]) -> FeedbackData:
        """Create new feedback in nps_feedback table."""
        try:
            if isinstance(feedback_data, CreateFeedbackRequest):
                data = feedback_data.model_dump()
            else:
                data = feedback_data

            # Convert rating (0-100) to score (0-10) for NPS
            rating = data.get("rating", 50)
            score = round(rating / 10) if rating else 5

            # Determine NPS feedback type based on score
            # 9-10 = promoter, 7-8 = passive, 0-6 = detractor
            if score >= 9:
                nps_feedback_type = "promoter"
            elif score >= 7:
                nps_feedback_type = "passive"
            else:
                nps_feedback_type = "detractor"

            logger.info(f"[FEEDBACK REPOSITORY] Creating NPS feedback: rating={rating}, score={score}, type={nps_feedback_type}")

            query = """
                INSERT INTO nps_feedback (simulation_id, user_id, score, feedback_type, comments)
                VALUES ($1, $2, $3, $4, $5)
                RETURNING id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
            """

            record = await fetchone(
                query,
                data["simulation_id"],
                data["user_id"],
                score,
                nps_feedback_type,
                data.get("comments"),
            )
            result_data = record_to_dict(record)

            # Convert to FeedbackData format (mapping nps_feedback fields)
            return FeedbackData(
                id=result_data["id"],
                simulation_id=result_data["simulation_id"],
                user_id=result_data["user_id"],
                competency_id=None,
                rating=score * 10,  # Convert back to 0-100 scale
                comments=result_data.get("comments"),
                feedback_type=result_data["feedback_type"],
                metadata=None,
                created_at=result_data.get("submitted_at"),
                updated_at=result_data.get("submitted_at"),
            )
        except Exception as e:
            logger.error(f"Error creating feedback: {e}")
            raise

    async def update(self, feedback_id: str, feedback_data: UpdateFeedbackRequest | dict[str, Any]) -> FeedbackData:
        """Update feedback."""
        try:
            if isinstance(feedback_data, UpdateFeedbackRequest):
                data = feedback_data.model_dump(exclude_unset=True)
            else:
                data = {k: v for k, v in feedback_data.items() if v is not None}

            # Build dynamic update query
            update_parts = []
            params = [feedback_id]
            param_idx = 2

            for key, value in data.items():
                if key == "metadata" and value and not isinstance(value, str):
                    value = json.dumps(value)
                update_parts.append(f"{key} = ${param_idx}")
                params.append(value)
                param_idx += 1

            if not update_parts:
                return await self.find_by_id(feedback_id)

            update_parts.append("updated_at = NOW()")
            query = f"""
                UPDATE feedback SET {', '.join(update_parts)}
                WHERE id = $1
                RETURNING id, simulation_id, user_id, competency_id, rating, comments, feedback_type, metadata, created_at, updated_at
            """
            record = await fetchone(query, *params)
            if not record:
                raise ValueError("Feedback not found")
            result_data = record_to_dict(record)
            if result_data.get("metadata") and isinstance(result_data["metadata"], str):
                try:
                    result_data["metadata"] = json.loads(result_data["metadata"])
                except json.JSONDecodeError:
                    pass
            return FeedbackData(**result_data)
        except Exception as e:
            logger.error(f"Error updating feedback: {e}")
            raise

    async def delete(self, feedback_id: str) -> None:
        """Delete feedback."""
        try:
            query = "DELETE FROM feedback WHERE id = $1"
            await execute(query, feedback_id)
        except Exception as e:
            logger.error(f"Error deleting feedback: {e}")
            raise

    async def delete_by_simulation_id(self, simulation_id: str) -> None:
        """Delete all feedback for a simulation."""
        try:
            query = "DELETE FROM feedback WHERE simulation_id = $1"
            await execute(query, simulation_id)
        except Exception as e:
            logger.error(f"Error deleting feedback by simulation ID: {e}")
            raise

    async def get_all_nps(self) -> list[NPSData]:
        """Get all NPS feedback."""
        try:
            query = """
                SELECT id, user_id, simulation_id, score, feedback_text, categories, follow_up_consent, submit_time_ms, created_at, updated_at
                FROM nps_feedback
                ORDER BY created_at DESC
            """
            records = await fetch(query)
            nps_list = []
            for record in records:
                data = dict(record)
                if data.get("categories") and isinstance(data["categories"], str):
                    try:
                        data["categories"] = json.loads(data["categories"])
                    except json.JSONDecodeError:
                        pass
                nps_list.append(NPSData(**data))
            return nps_list
        except Exception as e:
            logger.error(f"Error fetching all NPS: {e}")
            raise

    async def create_nps(self, nps_data: dict[str, Any]) -> NPSData:
        """Create NPS feedback."""
        try:
            nps_id = str(uuid4())
            query = """
                INSERT INTO nps_feedback (id, user_id, simulation_id, score, feedback_text, categories, follow_up_consent, submit_time_ms)
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
                RETURNING id, user_id, simulation_id, score, feedback_text, categories, follow_up_consent, submit_time_ms, created_at, updated_at
            """
            categories = nps_data.get("categories")
            if categories and not isinstance(categories, str):
                categories = json.dumps(categories)

            record = await fetchone(
                query,
                nps_id,
                nps_data["user_id"],
                nps_data.get("simulation_id"),
                nps_data["score"],
                nps_data.get("feedback_text"),
                categories,
                nps_data.get("follow_up_consent", False),
                nps_data.get("submit_time_ms"),
            )
            result_data = record_to_dict(record)
            if result_data.get("categories") and isinstance(result_data["categories"], str):
                try:
                    result_data["categories"] = json.loads(result_data["categories"])
                except json.JSONDecodeError:
                    pass
            return NPSData(**result_data)
        except Exception as e:
            logger.error(f"Error creating NPS feedback: {e}")
            raise


# Singleton instance
feedback_repository = FeedbackRepository()

File: backend/db/repositories/feedback-repository.ts
/**
 * Feedback Repository
 * Data access layer for feedback operations
 * Now uses nps_feedback table with NPS scoring system
 */

import { sql } from '../connection.ts';

export interface FeedbackData {
  id: string;
  simulation_id: string;
  user_id: string;
  competency_id?: string; // Not stored in nps_feedback, kept for API compatibility
  rating?: number; // 0-100 scale (converted from score)
  score?: number; // 0-10 scale (actual DB column)
  comments?: string;
  feedback_type: 'promoter' | 'passive' | 'detractor' | 'ai_generated' | 'user_submitted' | 'peer_review';
  created_at?: Date; // Mapped from submitted_at
  submitted_at?: Date; // Actual DB column
  updated_at?: Date; // Not in nps_feedback, kept for API compatibility
  reasons?: string[]; // NPS-specific field
}

export interface CreateFeedbackRequest {
  simulation_id: string;
  user_id: string;
  competency_id?: string; // Ignored for nps_feedback
  rating?: number; // 0-100 scale, will be converted to 0-10 score
  comments?: string;
  feedback_type: 'ai_generated' | 'user_submitted' | 'peer_review';
  reasons?: string[]; // Optional NPS reasons
}

export interface UpdateFeedbackRequest {
  rating?: number; // 0-100 scale
  comments?: string;
}

/**
 * Helper function to convert 0-100 rating to 0-10 NPS score
 */
function ratingToScore(rating: number): number {
  return Math.round(rating / 10);
}

/**
 * Helper function to convert 0-10 NPS score to 0-100 rating
 */
function scoreToRating(score: number): number {
  return score * 10;
}

/**
 * Helper function to determine NPS feedback type from score
 * 9-10 = Promoter, 7-8 = Passive, 0-6 = Detractor
 */
function scoreToFeedbackType(score: number): 'promoter' | 'passive' | 'detractor' {
  if (score >= 9) return 'promoter';
  if (score >= 7) return 'passive';
  return 'detractor';
}

export class FeedbackRepository {
  /**
   * Find feedback by ID
   */
  async findById(id: string): Promise<FeedbackData | null> {
    try {
      const result = await sql`
        SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
        FROM nps_feedback
        WHERE id = ${id}
      `;
      if (!result[0]) return null;

      // Convert to API format
      return {
        ...result[0],
        rating: scoreToRating(result[0].score),
        created_at: result[0].submitted_at,
      };
    } catch (error) {
      console.error('Error finding feedback by ID:', error);
      throw error;
    }
  }

  /**
   * Get feedback by simulation ID
   */
  async findBySimulationId(simulationId: string): Promise<FeedbackData[]> {
    try {
      const result = await sql`
        SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
        FROM nps_feedback
        WHERE simulation_id = ${simulationId}
        ORDER BY submitted_at DESC
      `;

      // Convert to API format
      return result.map((row: any) => ({
        ...row,
        rating: scoreToRating(row.score),
        created_at: row.submitted_at,
      }));
    } catch (error) {
      console.error('Error fetching feedback by simulation:', error);
      throw error;
    }
  }

  /**
   * Get feedback by user ID
   */
  async findByUserId(userId: string): Promise<FeedbackData[]> {
    try {
      const result = await sql`
        SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
        FROM nps_feedback
        WHERE user_id = ${userId}
        ORDER BY submitted_at DESC
      `;

      // Convert to API format
      return result.map((row: any) => ({
        ...row,
        rating: scoreToRating(row.score),
        created_at: row.submitted_at,
      }));
    } catch (error) {
      console.error('Error fetching feedback by user:', error);
      throw error;
    }
  }

  /**
   * Get feedback by competency ID
   * Note: nps_feedback table doesn't have competency_id, returns empty array
   */
  async findByCompetencyId(competencyId: string): Promise<FeedbackData[]> {
    try {
      console.warn('[FEEDBACK REPOSITORY] findByCompetencyId called but nps_feedback table does not have competency_id column');
      return [];
    } catch (error) {
      console.error('Error fetching feedback by competency:', error);
      throw error;
    }
  }

  /**
   * Get feedback by type (NPS types: promoter, passive, detractor)
   */
  async findByType(type: 'ai_generated' | 'user_submitted' | 'peer_review' | 'promoter' | 'passive' | 'detractor'): Promise<FeedbackData[]> {
    try {
      const result = await sql`
        SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
        FROM nps_feedback
        WHERE feedback_type = ${type}
        ORDER BY submitted_at DESC
      `;

      // Convert to API format
      return result.map((row: any) => ({
        ...row,
        rating: scoreToRating(row.score),
        created_at: row.submitted_at,
      }));
    } catch (error) {
      console.error('Error fetching feedback by type:', error);
      throw error;
    }
  }

  /**
   * Get all feedback
   */
  async findAll(limit?: number): Promise<FeedbackData[]> {
    try {
      const query = limit
        ? sql`SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
             FROM nps_feedback
             ORDER BY submitted_at DESC
             LIMIT ${limit}`
        : sql`SELECT id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
             FROM nps_feedback
             ORDER BY submitted_at DESC`;

      const result = await query;

      // Convert to API format
      return result.map((row: any) => ({
        ...row,
        rating: scoreToRating(row.score),
        created_at: row.submitted_at,
      }));
    } catch (error) {
      console.error('Error fetching all feedback:', error);
      throw error;
    }
  }

  /**
   * Create feedback
   * Converts 0-100 rating to 0-10 NPS score and determines feedback type
   */
  async create(feedbackData: CreateFeedbackRequest): Promise<FeedbackData> {
    try {
      // Convert rating to score (0-100 -> 0-10)
      const score = feedbackData.rating ? ratingToScore(feedbackData.rating) : 5;

      // Determine NPS feedback type based on score
      const npsFeedbackType = scoreToFeedbackType(score);

      console.log(`[FEEDBACK REPOSITORY] Creating NPS feedback: rating=${feedbackData.rating}, score=${score}, type=${npsFeedbackType}`);

      const result = await sql`
        INSERT INTO nps_feedback (
          simulation_id, user_id, score, feedback_type, reasons, comments
        )
        VALUES (
          ${feedbackData.simulation_id},
          ${feedbackData.user_id},
          ${score},
          ${npsFeedbackType},
          ${feedbackData.reasons || null},
          ${feedbackData.comments || null}
        )
        RETURNING id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
      `;

      // Convert to API format
      return {
        ...result[0],
        rating: scoreToRating(result[0].score),
        created_at: result[0].submitted_at,
      };
    } catch (error) {
      console.error('Error creating feedback:', error);
      throw error;
    }
  }

  /**
   * Update feedback
   * Converts rating to score and recalculates feedback type
   */
  async update(id: string, feedbackData: UpdateFeedbackRequest): Promise<FeedbackData> {
    try {
      const updates: string[] = [];

      if (feedbackData.rating !== undefined) {
        const score = ratingToScore(feedbackData.rating);
        const npsFeedbackType = scoreToFeedbackType(score);
        updates.push(`score = ${score}`);
        updates.push(`feedback_type = '${npsFeedbackType}'`);
      }
      if (feedbackData.comments !== undefined) {
        updates.push(`comments = '${feedbackData.comments.replace(/'/g, "''")}'`); // Escape single quotes
      }

      if (updates.length === 0) {
        throw new Error('No fields to update');
      }

      const result = await sql`
        UPDATE nps_feedback
        SET ${sql.raw(updates.join(', '))}
        WHERE id = ${id}
        RETURNING id, simulation_id, user_id, score, feedback_type, reasons, comments, submitted_at
      `;

      // Convert to API format
      return {
        ...result[0],
        rating: scoreToRating(result[0].score),
        created_at: result[0].submitted_at,
      };
    } catch (error) {
      console.error('Error updating feedback:', error);
      throw error;
    }
  }

  /**
   * Delete feedback
   */
  async delete(id: string): Promise<void> {
    try {
      await sql`
        DELETE FROM nps_feedback
        WHERE id = ${id}
      `;
    } catch (error) {
      console.error('Error deleting feedback:', error);
      throw error;
    }
  }

  /**
   * Delete feedback by simulation ID
   */
  async deleteBySimulationId(simulationId: string): Promise<void> {
    try {
      await sql`
        DELETE FROM nps_feedback
        WHERE simulation_id = ${simulationId}
      `;
    } catch (error) {
      console.error('Error deleting feedback by simulation:', error);
      throw error;
    }
  }

  /**
   * Get average rating by competency
   * Note: nps_feedback table doesn't have competency_id, returns 0
   */
  async getAverageRatingByCompetency(competencyId: string): Promise<number> {
    try {
      console.warn('[FEEDBACK REPOSITORY] getAverageRatingByCompetency called but nps_feedback table does not have competency_id column');
      return 0;
    } catch (error) {
      console.error('Error calculating average rating:', error);
      throw error;
    }
  }

  /**
   * Get user feedback statistics
   */
  async getUserFeedbackStats(userId: string): Promise<{
    total: number;
    avgRating: number;
    byType: Record<string, number>;
  }> {
    try {
      const totalResult = await sql`
        SELECT COUNT(*) as total, AVG(score) as avg_score
        FROM nps_feedback
        WHERE user_id = ${userId}
      `;

      const byTypeResult = await sql`
        SELECT feedback_type, COUNT(*) as count
        FROM nps_feedback
        WHERE user_id = ${userId}
        GROUP BY feedback_type
      `;

      const byType: Record<string, number> = {};
      byTypeResult.forEach((row: any) => {
        byType[row.feedback_type] = parseInt(row.count);
      });

      // Convert average score to rating (0-10 -> 0-100)
      const avgScore = parseFloat(totalResult[0].avg_score) || 0;
      const avgRating = scoreToRating(avgScore);

      return {
        total: parseInt(totalResult[0].total),
        avgRating,
        byType,
      };
    } catch (error) {
      console.error('Error fetching user feedback stats:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const feedbackRepository = new FeedbackRepository();
